controllers/machinehealthcheck_controller.go

```diff
@@ -0,0 +1,135 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// MachineHealthCheckReconciler reconciles a MachineHealthCheck object
+type MachineHealthCheckReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	controller controller.Controller
+	recorder   record.EventRecorder
+}
+
+func (r *MachineHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineHealthCheck{}).
+		WithOptions(options).
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.controller = controller
+	r.recorder = mgr.GetEventRecorderFor("machinehealthcheck-controller")
+	return nil
+}
+
+func (r *MachineHealthCheckReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+	logger := r.Log.WithValues("machinehealthcheck", req.Name, "namespace", req.Namespace)
+
+	// Fetch the MachineHealthCheck instance
+	m := &clusterv1.MachineHealthCheck{}
+	if err := r.Client.Get(ctx, req.NamespacedName, m); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, m.ObjectMeta.Namespace, m.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get cluster %q for machine %q in namespace %q",
+			m.Spec.ClusterName, m.Name, m.Namespace)
+	}
+
+	// Return early if the object or Cluster is paused.
+	if util.IsPaused(cluster, m) {
+		logger.V(3).Info("reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(m, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, m); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Reconcile labels.
+	if m.Labels == nil {
+		m.Labels = make(map[string]string)
+	}
+	m.Labels[clusterv1.ClusterLabelName] = m.Spec.ClusterName
+
+	result, err := r.reconcile(ctx, cluster, m)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile MachineHealthCheck")
+		r.recorder.Eventf(m, corev1.EventTypeWarning, "ReconcileError", "%v", err)
```

Comment:
It looks like this is currently swallowing the error, shouldn't this return the error instead?

Reply:
Probably depends more on what the actual impl of reconcile() does and what an error means (should we retry or not)

Reply:
What is the behavior we expect if the reconciliation fails?

Do we want to retry immediately (requeue option is set in the result, or an error is returned), or do we want to wait a bit and retry again after a short wait (requeue and requeueAfter options are set in the result, no error returned), or do we just want to wait until something else triggers a reconciliation such as a watched resource is updated or full reconciliation is triggered by the informer (no error, no requeue option set in result)?

For the retry after a certain period, we generally handle that in a way similar to: https://github.com/kubernetes-sigs/cluster-api/blob/76ea21f93aaea325f5e44e815e364fc02b3008b0/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go#L215-L221

Reply:
I think for the moment, this isn't actually very obvious. I suspect this will become more obvious once the next PR comes up and we can see what the actual errors from the inner reconcile are. My suggestion would be, to for now, have this log the error but not return anything (so we aren't continuously reconciling a "not implemented" error) and then add a TODO to revisit this in the next PR once we can see what the behaviour we want is? WDYT?

Reply:
sgtm